Add the ability to output vmdk via qemu-img

jpculp opened the pull request:
**Issue number:**
N/A

**Description of changes:**
This change allows users to edit the metadata in a variant's cargo.toml to output an OVA compatible vmdk through buildsys via qemu-img.

Double quotes were also added to several places in rpm2img to prevent unwanted globbing and word splitting (SC2086).


**Testing done:**

- Built aws-k8s-1.18 variant with no toml changes and received the standard img.lz4 output.

- Built aws-k8s-1.18 variant with ```image-format = "vmdk"``` added to the toml and received stream-optimized vmdk output for bottlerocket-aws-k8s-1.18-aarch64 and bottlerocket-aws-k8s-1.18-aarch64-data.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
Add the ability to output vmdk via qemu-img

This change allows users to edit the metadata in a variant's cargo.toml
to output a vmdk through buildsys via qemu-img.

jpculp commented on Makefile.toml:
```diff
@@ -381,18 +381,11 @@ dependencies = ["build-variant"]
-  ${BUILDSYS_OUTPUT_DIR}/latest/${BUILDSYS_NAME_VARIANT}-root.verity.lz4
-ln -snf ../${BUILDSYS_NAME_FULL}-migrations.tar \
-  ${BUILDSYS_OUTPUT_DIR}/latest/${BUILDSYS_NAME_VARIANT}-migrations.tar
+for artifact in ${BUILDSYS_OUTPUT_DIR}/${BUILDSYS_NAME_FULL}*; do
```
One pitfall of using a loop here is if a user builds a variant as an image, and then later rebuilds that same variant as a vmdk, if the build-id does not change it will link the old artifacts img.lz4 as well.

jahkeup replied:
That might be desirable in some situations, for example: to build both a raw disk image *and* a VMDK.

bcressey replied:
I have a fix coming soon for #1027 that should address this; artifacts from a previous `buildsys` invocation will be cleaned up prior to the next, so we shouldn't hit the situation where we have both types at once.

zmrow replied:
@jpculp Let's remove this loop.  If another solution is coming we don't need to fix the "problem" here.

jpculp replied:
@bcressey  can correct me, but I think his solution is just for cleaning up old artifacts, such as in the situation I mentioned above. This loop would still be necessary for generating the proper symlinks under the build/images/latest directory.

zmrow replied:
Ah yeah you're right!

jahkeup commented on tools/buildsys/src/builder.rs:
```diff
@@ -89,6 +95,19 @@ impl VariantBuilder {
+            Some(ImageFormat::VMDK) => String::from("vmdk"),
+        };
+        let image_options = match image_options {
+            Some(options_string) => options_string.join(","),
```
nit: `options_string` isn't a string here, let's avoid suggesting it is:
```suggestion
            Some(opts) => opts.join(","),
```

jpculp replied:
Makes sense to me, but it looks like we're just going to remove image_options for now.

jahkeup commented on tools/buildsys/src/manifest.rs:
```diff
@@ -152,6 +162,15 @@ pub(crate) struct BuildPackage {
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "lowercase")]
+pub(crate) enum ImageFormat {
+    IMG,
```
Can we refer to `IMG` instead as `Raw`? The sort of files that `IMG` is naming are typically called `raw` by tools and the various virtualization communities.

bcressey commented on tools/rpm2img:
```diff
@@ -90,43 +95,43 @@ rm -rf "${ROOT_MOUNT}"/var/lib "${ROOT_MOUNT}"/usr/share/licenses/*
-echo "PRETTY_NAME=\"Bottlerocket OS ${VERSION_ID}\"" >> ${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release
-echo "VARIANT_ID=${VARIANT}" >> ${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release
-echo "VERSION_ID=${VERSION_ID}" >> ${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release
-echo "BUILD_ID=${BUILD_ID}" >> ${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release
+echo "PRETTY_NAME=\"Bottlerocket OS ${VERSION_ID}\"" >> "${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release"
+echo "VARIANT_ID=${VARIANT}" >> "${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release"
+echo "VERSION_ID=${VERSION_ID}" >> "${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release"
+echo "BUILD_ID=${BUILD_ID}" >> "${ROOT_MOUNT}/${SYS_ROOT}/usr/lib/os-release"
```
I'd like to have all these unrelated shellcheck fixes in a separate commit.

bcressey commented on tools/buildsys/src/builder.rs:
```diff
@@ -89,6 +95,19 @@ impl VariantBuilder {
+            Some(ImageFormat::IMG) | None => String::from("img"),
+            Some(ImageFormat::VMDK) => String::from("vmdk"),
+        };
+        let image_options = match image_options {
+            Some(options_string) => options_string.join(","),
```
`image_opts` feels like the wrong abstraction boundary here - it's not consistent with the other handling of low-level details that takes place in `rpm2img`.

Let's hardcode the stream-optimized options in `rpm2img` when building VMDKs. We can name the format `stream-vmdk` if we need to distinguish it from ordinary VMDK for some reason, but I don't think we do.

zmrow commented on tools/rpm2img:
```diff
@@ -3,22 +3,26 @@
 shopt -qs failglob
 
+OUTPUT_FMT="raw"
+OUTPUT_OPTS="none"
```
Let's eliminate this option for now

zmrow commented on tools/rpm2img:
```diff
@@ -206,16 +210,34 @@ dd if="${BOTTLEROCKET_DATA}" of="${DATA_IMAGE}" conv=notrunc bs=1M seek=1
+  fi
+
+  if [[ ${OUTPUT_OPTS} != "none" ]]; then
+    QEMU_PARAMS="${QEMU_PARAMS} -o ${OUTPUT_OPTS}"
```
I think this is a bit over complicated.  We can simplify it a bit by eliminating the multiple extra variables and having the qemu command hardcoded for now.  If you want to encapsulate all qemu args in a single variable that can be re-used, that's fine too.

IMHO it also might help to match on a positive case first to make it simpler to read:

```
if [[ ${OUTPUT_FMT} == "vmdk" ]]; then
    # Don't need to use this variable, could just hard code the commands
    QEMU_OPTS="convert -f raw -O vmdk -o subformat=streamOptimized"

    qemu-img "${QEMU_OPTS}" .....

# Can use elif's later for additional output formats
else
    ....
fi
```__

jpculp replied:
I've updated my commits with something in-between your suggestion and the original commit.

zmrow reviewed (approved):
:hamburger:

bcressey commented on tools/rpm2img:
```diff
@@ -3,22 +3,25 @@
-VERITY_IMAGE_NAME="${FILENAME_PREFIX}-root.verity.lz4"
-ROOT_IMAGE_NAME="${FILENAME_PREFIX}-root.ext4.lz4"
-DATA_IMAGE_NAME="${FILENAME_PREFIX}-data.img.lz4"
+DISK_IMAGE_BASENAME="${FILENAME_PREFIX}"
+BOOT_IMAGE_BASENAME="${FILENAME_PREFIX}-boot"
+VERITY_IMAGE_BASENAME="${FILENAME_PREFIX}-root"
+ROOT_IMAGE_BASENAME="${FILENAME_PREFIX}-root"
+DATA_IMAGE_BASENAME="${FILENAME_PREFIX}-data"
```
It's weird to me that `VERITY_IMAGE_BASENAME` and `ROOT_IMAGE_BASENAME` have the same value. Maybe incorporate part of the extension?

```suggestion
DISK_IMAGE_BASENAME="${FILENAME_PREFIX}"
DATA_IMAGE_BASENAME="${FILENAME_PREFIX}-data"

BOOT_IMAGE_BASENAME="${FILENAME_PREFIX}-boot.ext4"
ROOT_IMAGE_BASENAME="${FILENAME_PREFIX}-root.ext4"
VERITY_IMAGE_BASENAME="${FILENAME_PREFIX}-root.verity"
```

Then the final extension (`.img.lz4`, `.vmdk`, `.lz4`) would be provided on the same line where we invoke the tool that converts it to that format, which would keep the clarity improvement you're after.

jpculp replied:
I can see how that would be strange. I've gone ahead and amended the commit so that we only use BASENAME for the disk and data images.

bcressey commented on tools/rpm2img:
```diff
@@ -206,14 +211,28 @@ dd if="${BOTTLEROCKET_DATA}" of="${DATA_IMAGE}" conv=notrunc bs=1M seek=1
+  qemu-img convert -f raw -O vmdk -o subformat=streamOptimized "${DATA_IMAGE}" "${OUTPUT_DIR}/${DATA_IMAGE_BASENAME}.vmdk"
+  chown 1000:1000 "${OUTPUT_DIR}/${DISK_IMAGE_BASENAME}.vmdk" \
+    "${OUTPUT_DIR}/${DATA_IMAGE_BASENAME}.vmdk"
+else
+  qemu-img convert -f raw -O "${OUTPUT_FMT}" "${DISK_IMAGE}" "${OUTPUT_DIR}/${DISK_IMAGE_BASENAME}.${OUTPUT_FMT}"
+  qemu-img convert -f raw -O "${OUTPUT_FMT}" "${DATA_IMAGE}" "${OUTPUT_DIR}/${DATA_IMAGE_BASENAME}.${OUTPUT_FMT}"
+  chown 1000:1000 "${OUTPUT_DIR}/${DISK_IMAGE_BASENAME}.${OUTPUT_FMT}" \
+    "${OUTPUT_DIR}/${DATA_IMAGE_BASENAME}.${OUTPUT_FMT}"
+fi
```
If we don't currently expect to take this path, I'd prefer to just error out:

```suggestion
else
  echo "unexpected image format '${OUTPUT_FMT}'" >&2
  exit 1
fi
```

jpculp replied:
After some further discussion, I removed the else entirely and added a case statement to check for valid output formats immediately after the argument parsing. To support additional formats in the future, they will have to be added to the case statement and an additional elif will have to be appended to this block.

bcressey reviewed (commented):
LGTM, just one small concern.

jpculp committed:
Add double quotes to some unwrapped variables

Double quotes were added to several places in rpm2img to further prevent
unwanted globbing and word splitting (SC2086).

zmrow reviewed (approved):
:medal_military:
